PHI 11 - Type Constructors

hejohns opened the pull request.

bkase committed:
Initial draft

Merges in content from PHI 5 (type variables) with latex judgements
for bidirectional type elaboration and kind synthesis/analysis. Will
follow up with content from the PHI 3 (type aliases) later.

bkase committed:
Adds note about type holes

bkase committed:
Delta supports htyp/varctx or kind/tyvarctx

bkase committed:
Adds kind assignment, fixes a few bugs

bkase committed:
Updated for kind-assignment and sane delta

bkase committed:
Uses cdot for both empty sets

bkase committed:
Attempts to add Ana uvar rule

bkase committed:
Judgements and changes for singleton kinds

bkase committed:
Requested changes

bkase committed:
Fixes issues in PHI, adds theorems for reasoning

bkase committed:
Unrecognition tweak

bkase committed:
Fixes for broken handling of tyvars

bkase committed:
Made suggested changes

bkase committed:
Latest changes

bkase committed:
Moves to bidirectional kind assignment

bkase committed:
Tweaks to design docs

bkase committed:
Removes the extra recursive param in lines

bkase committed:
Remove kind ascription on type equivalence

bkase committed:
Synthesize self for tvars

hejohns committed:
htau should be dtau in judgment box

hejohns committed:
KSVar bug?

hejohns committed:
let's backup before any possible hard drive failures!

hejohns committed:
ditto

hejohns committed:
good place to stop

need to make a TypeEquiv^ and a TypeEquivv

hejohns committed:
forgot some stuff

hejohns committed:
small fixes before reading

hejohns committed:
circularity mostly worked out?

except the rules are obnoxiously opaque and derivations tediously convuluted

hejohns committed:
progress. I think

hejohns committed:
::>, :: better, but equivalence at superkinds still broken

hejohns committed:
is this necessary for consistentsubkinding+labeled singletons?

hejohns committed:
I think the solid rules are finally right.......

hejohns committed:
change ordering

hejohns committed:
KindWellFormed

hejohns committed:
auto number rules

hejohns committed:
mucking around

hejohns committed:
probably should've named the branch 11-parameterized-types

hejohns committed:
rename things

hejohns committed:
does 21 make sense??

hejohns committed:
a bunch of well formedness stuff

hejohns committed:
sad. needs "redundant" premiss to be provable

hejohns committed:
slow progess

hejohns committed:
please don't let weaking be sim. inductive as well :(

hejohns committed:
remove redundant premisses

also we need to start naming rules. When the numbers change as rules get
moved or deleted, it's really annoying

hejohns committed:
?! looks strange. But might actually work

hejohns committed:
are holes handled correctly now??

hejohns committed:
title

hejohns committed:
equality of types and KHole strangeness

hejohns committed:
type equality fixed?(take 2)

hejohns committed:
actually need this rule after all

for when eg type function is a variable, so can't look inside body with
normalization/reduction

can probably prove transativity etc of type equality as a consequence of
transativity etc of kind equality, but will be some work for sure

hejohns committed:
yeah ok we do need refl,symm,trans for type equality

otherwise we wouldn't be able to show the two ways to reduce a singleton
of a singleton are equivalent

hejohns committed:
naming rules

hejohns committed:
about to regex so let's backup

hejohns committed:
eh

hejohns committed:
all rules named

hejohns committed:
weakening

hejohns committed:
idk why I insisted on typing it all out last time

hejohns committed:
fix exchange

hejohns committed:
I screwed up the layout of these lemmas

hejohns committed:
huge waste of a day. but at least it worked out in the end?

hejohns committed:
fix context marking

hejohns committed:
:(

hejohns committed:
dodged a KOMA-Script bullet

hejohns committed:
:(

hejohns committed:
:-()

hejohns committed:
I sure hope the trees swapped w/o error

hejohns committed:
I hate myself for not reversing the inferrule args earlier

well I guess we're never editing the Weakening proof

hejohns committed:
got vim macro'd to hell. We'll fix the spacing another time

hejohns committed:
we've gone back/forth 100 times but COK does need subderivation

hejohns committed:
ended up just fixing weakening manually. had to fix an error

hejohns committed:
this error

hejohns committed:
weaking looks good!

hejohns committed:
finally figured out how to use typearea

hejohns committed:
I think we need induction on type size for K-subst

hejohns committed:
lemma -> theorem, OK-PK

hejohns committed:
must've changed DIV on accident

hejohns committed:
meant THIS one

hejohns committed:
sending it out to cyrus-

hejohns committed:
might as well add the pptx

hejohns committed:
fell asleep; saving

hejohns committed:
singleton reduction/normalization rules?

no idea if this is what was intended at the end of summer-- will check
sometime in the future

hejohns committed:
named singleton reduction rules

use KindWellFormed for premiss of \SReducOp\hyph 1

hejohns committed:
mostly sketch PK-Unicity

surprisingly, "mn-nicity" for singleton reduction isn't trivial to show
(or I'm just slow. I'll use the time as an excuse)

hejohns committed:
singleton normalization unicity took too many steps

hejohns committed:
tiny bit of notes to myself

hejohns committed:
syntastic

hejohns committed:
we actually do need this rule

type equivalence at reasonable non-Type kinds should also become
provable. Rule does look slightly strange though.

Note to self:
A similar rule does not appear in Stone/Harper 2006 since their
term-type system does not have an analogue to arrow/product/sum types

There would have to be some term like `e1 <*> e2 : b if e1 : b and e2 :
b` eg if the only base type is `int`, then `<*>` as a primitive addition

hejohns committed:
put code scratch here

hejohns committed:
HUnit instead

hejohns committed:
The most trivial cases

utf8 ftw

hejohns committed:
what a waste of time fiddling w/ (.)/flip...

hejohns committed:
finish αRename

also we're using hindent now!
The broken autoindent (well, broken on haskell, works fine for C) was
driving me crazy

hejohns committed:
canonical form stuff

hejohns committed:
just to save what's in Algo so far

hejohns committed:
hindent/various cosmetics

hejohns committed:
too tired, just add everything

hejohns committed:
wrap up last commit sorta-- canon isn't totally broken

hejohns committed:
move canon to Algo

why was it even in Ctx? Makes no sense

hejohns committed:
Need pk wfak csk now

hejohns committed:
need to do csk, then we can test some stuff

hejohns committed:
hs-perl5 actually works.......

hejohns committed:
regoranize tests to make running only some tests easier

(some tests were causing looping and others were spamming the console
 with errors and undefined s)

hejohns committed:
what a fun way to implement fresh

hejohns committed:
focusing on getting tests working

hejohns committed:
working on very basic test. Still not sure how to handle the mutually recursive judgments

hejohns committed:
some self-induced pain because I didn't want to use de bruijn indeces

hejohns committed:
≡_α for Typ s (although idk where we need it) as part of Rewrite class

hejohns committed:
Good canonKnd test

hejohns committed:
debugging trace calls

(forgot why the short circuit was needed)

hejohns committed:
current tests pass now; can keep moving forward

hejohns committed:
check ≡_α in csk

hejohns committed:
use kequiv instead of just ≡_α where appropriate

hejohns committed:
some basic tests pass

hejohns committed:
some documentation for myself (I think I'll Reason this soon)

hejohns committed:
need to run; doesn't build; need to handle base types like `list`

hejohns committed:
partial revert (working on η rules)

hejohns committed:
apparently hindent really doesn't like deriving via

hejohns committed:
forgot haskell has type families. doesn't build-- lots of changes

hejohns committed:
I didn't think this through-- need to revert last couple commits

way to spend ~2 hours coding something I forgot I didn't think through
fully cause I'm tired and going on autopilot
done for today

hejohns committed:
revert most of last couple commits (see last commit)

(does not build)

hejohns committed:
clean the crumbs from last revert-- make it build

hejohns committed:
need to run, but our current plan is to have two Aps for irriducible/reducible

hejohns committed:
leaving this here, but we're gonna branch and move a lot

hejohns committed:
idk what's what anymore-- adding and then creating new files

hejohns committed:
reorganizing everything

hejohns committed:
lots of changes. Reorganized everything to ex/internal language

Hung up on part of elaboration.

hejohns committed:
tooling paaaiiinn

I need to vent
hindent's parser can't handle record dot syntax
stylish-haskell's can, but stylish-haskell just doesn't really work for
some reason

but I need record dot syntax to use duplicate record fields this way

arghhh

hejohns committed:
elab

hejohns committed:
a little wh/normal + hindent

hejohns committed:
path normalization reduces in the kind-- need to work out

hejohns committed:
tests pass again. But I'm not really sure what I've done

hejohns committed:
mpk

hejohns committed:
used wrong ctx function

hejohns committed:
test exception call stack

hejohns committed:
hmmm not fun

hejohns committed:
simple test

hejohns committed:
forgot recursive type_normal for HO singletons

hejohns committed:
parser for writing tests: SEE LONG MESSAGE

alex Lexer.x
happy -l Parser.y # the grammer prob sucks

parse :: String -> External.Typ
(I'll make another for kinds in a second)

hejohns committed:
some tests from before

hejohns committed:
well the test fails... but due to perl eval...

hejohns committed:
I cannot l;we;nasdk

hejohns committed:
adsf

hejohns committed:
epic fail. I mean seriously. ~5 hours wasted

And I've written this before.
Idk what to say

hejohns committed:
external expression langauge

hejohns committed:
Bidirectional typechecker basics

on simplified external expression langauge
no tests (only some repl sanity checking)
no expression holes

hejohns committed:
syn tests

hejohns committed:
Didn't realize I didn't add BiDirectional.hs ... rip git history

re Friday's chat, added a couple parametized type tests and ran into
some nonobvious bugs (of course...)
so there are only a couple more tests-- they took a little bit to fix

hejohns committed:
writing up algorithmic rules

latex key/value is so much more painful than I could've ever anticipated

hejohns committed:
type normal

the last rule needs to get checked-- too tired. Just mechanically wrote
it

hejohns committed:
did that perl filter actually work

hejohns committed:
AlgTypeNormal

hejohns committed:
AlgWHNormal

hejohns committed:
make sure binop types normalize correctly

hejohns committed:
AlgNatKind

hejohns committed:
typo in path normal (caught during latexing). TODO: test should ve exposed

hejohns committed:
AlgPathNormal

hejohns committed:
ElimCtx not EvalCtx

hejohns committed:
number alg rules for now

now there's \rule, \rinfer, \rrule, \inferrule...

hejohns committed:
start gradualizing type equality

hejohns committed:
some perl pie action: |- to ⊢

hejohns committed:
typo

hejohns committed:
term_normal at KHole?

hejohns committed:
initial gradualization and csk and kequiv

hejohns committed:
I think caught up gradual

too tired to check

hejohns committed:
as discussed in OH minus TODO s

hejohns committed:
part way through tidying up for PR

fiddled around w/ latex for a long time
this doc is really quite a mess

hejohns committed:
make isPath a judgment

hejohns committed:
partial revert: we did need this whreduc rule

hejohns committed:
minor fixes

hejohns committed:
clarify normalize at kind and fill in TODO

hejohns committed:
clean up notes

cyrus- commented on 11-type-constructors/11-type-constructors.md:
```diff
@@ -0,0 +1,24 @@
+
+## Introduction
+
+see my SURE poster for a motivating example.  
```
this still needs to be written

cyrus- commented on 11-type-constructors/latex/joshuadunfield.sty:
```diff
@@ -0,0 +1,699 @@
+\ProvidesClass{joshuadunfield}[2015/06/20 - Minor revision]
```
jana dunfield no longer goes by joshua so let's rename this style file

hejohns replied:
0832702f79d05851766f46d19edcc0b1a7b3860b

cyrus- commented on 11-type-constructors/latex/11-type-constructors.tex:
```diff
@@ -0,0 +1,920 @@
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```
there are two 11-type-constructors.tex files, one here and one in the parent directory. which one am I reviewing? please consolidate.

hejohns replied:
d549b0c72c3d907d22a7d90462ec39d6557f0f55

(I was using the upper .tex to generate the .md phi information)

hejohns committed:
rename md tex source to reduce confusion

hejohns committed:
rename j.....dunfield.sty by getting upstream version

I've been getting problems w/ \newcommand{\mathcolor}{..} redefinition
in this file recently (w/ the old 2015 version).
How did this ever compile? Something upstream with amsmath or something
must've changed since none of the previous commits build

hejohns committed:
add information to jdunfield.sty header

hejohns committed:
comment out \newcommand{\mathcolor}{..} for now and build

hejohns committed:
fix \today substitution

I swear it worked before though

hejohns committed:
update hejohns-colors

hejohns committed:
fixup latex, incorporating recent hejohns-stub changes and comment out unwanted

hejohns committed:
need to write md

cyrus- commented on 11-type-constructors/.11-type-constructors.tex:
```diff
@@ -0,0 +1,13 @@
+\documentclass[12pt]{article}
```
this file should be deleted I think.

hejohns replied:
It's to generate the markdown file
I can delete it since it seems like it's confusing

cyrus- commented on 11-type-constructors/algo/src/External.hs:
```diff
@@ -0,0 +1,99 @@
+  | Π TID Knd Knd
+  deriving (Show)
+
+-- I'm afraid that sooner or later I'll run into a case where just using De
```
why didn't you use de Bruijn indices? (that's what we're doing in the Hazel implementation)

hejohns replied:
At the time, I thought:
-- the phi uses variable names, so the implementation might as well too to make sure the renaming rules are right
-- it might be easier to debug w/o de Bruijn indices
-- I was going to use a Barendregt convention anyways, so it should be a little easier to implement
(of course the third point doesn't actually make sense, and is somewhat opposed to the first)

I agree in retrospect that I should've done de Bruijn indices like you'd said at the beginning, but I was still learning these things and wrote all the renaming stuff early on

Now there's not much point in changing it in the haskell implementation right?

cyrus- commented on 11-type-constructors/algo/src/ICtx.hs:
```diff
@@ -0,0 +1,69 @@
+  | Ctx :⌢⌢⌢ EAssump
+  deriving (Show)
+
+-- TODO: this is c/p from ECtx
```
what is this TODO saying? that the external and internal languages are the same?

hejohns replied:
yeah that the external and internal type level languages are the same, but the internal language should have kind casts

cyrus- commented on 11-type-constructors/algo/src/Internal.hs:
```diff
@@ -0,0 +1,94 @@
+-- The internal type language
+-- (following ``convention'', we speak of terms and types, not types and kinds
+-- in the context of normalization)
+-- TODO: In the future, should contain kind casts for a full gradual
```
what can't currently be expressed?

hejohns replied:
if you write something like
```
type T = Int in
type V = λt::KHole.t Bool in
type Bogus = V T in
...
```
kind casts would let you express why Bogus is bogus

cyrus- commented on 11-type-constructors/algo/test/Tests.hs:
```diff
@@ -0,0 +1,252 @@
+        False
+      , Nil ⊢ tequiv (Tλ "t" Type $ TVar "t") (Tλ "t" Type Bse) Type ~?= False
+      , Nil ⊢
+        tequiv (Tλ "t" Type $ TVar "t") (Tλ "t" Type Bse) (Π "t" Type Type) ~?=
```
"t" isn't in scope in the annotation on Tλ right?

hejohns replied:
Yes, although I'm confused what that has to do with those tests?

cyrus- commented on 11-type-constructors/algo/test/Tests.hs:
```diff
@@ -0,0 +1,252 @@
+-- tequiv τ1 τ2 κ = \aΓ -> Algo.tequiv aΓ τ1 τ2 κ
+-- since application precedence is so high (and can't be competed against)
+tequiv :: _
+tequiv = ((.) . (.) $ flip) . ((.) flip) . flip $ Algo.tequiv
```
uhhh

hejohns replied:
I just wanted to try to rearrange the arguments point free/with combinators. To see if I could do it.

cyrus- reviewed (changes requested):
Gave everything at least a quick look. Have you checked that the rules in the PDF still correspond to what is implemented in algo?

See comments otherwise. Do you think you'll have time soon to implement into Hazel now that type variables are imminent?

hejohns committed:
bump to newer ghc by removing unused Language.Perl dependency

hejohns committed:
untrack tex to generate md-- just don't accidentally rm it locally...

hejohns committed:
...and we're going back to older ghc since hls doesn't work for some reason

and I'm really not in the mood to try to fix it right now when I can't
even seem to build it from source and yes I'm annoyed right now
